sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceV2ReadSuite.scala

```diff
@@ -0,0 +1,503 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution.datasources.v2.state
+
+import org.scalatest.Assertions
+
+import org.apache.spark.sql.{AnalysisException, DataFrame, Encoders, Row}
+import org.apache.spark.sql.catalyst.expressions.{BoundReference, GenericInternalRow}
+import org.apache.spark.sql.catalyst.plans.physical.HashPartitioning
+import org.apache.spark.sql.execution.datasources.v2.state.utils.SchemaUtil
+import org.apache.spark.sql.execution.streaming.MemoryStream
+import org.apache.spark.sql.execution.streaming.state.{HDFSBackedStateStoreProvider, RocksDBStateStoreProvider}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.streaming.OutputMode
+import org.apache.spark.sql.types.{IntegerType, StructType}
+
+class HDFSBackedStateDataSourceV2ReadSuite extends StateDataSourceV2ReadSuite {
+  private var oldProviderClass: String = _
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+
+    oldProviderClass = spark.conf.get(SQLConf.STATE_STORE_PROVIDER_CLASS.key)
+    spark.conf.set(SQLConf.STATE_STORE_PROVIDER_CLASS.key,
+      classOf[HDFSBackedStateStoreProvider].getName)
+  }
+
+  override def afterAll(): Unit = {
+    if (oldProviderClass != null) {
+      spark.conf.set(SQLConf.STATE_STORE_PROVIDER_CLASS.key, oldProviderClass)
+    } else {
+      spark.conf.unset(SQLConf.STATE_STORE_PROVIDER_CLASS.key)
+    }
+
+    super.afterAll()
+  }
+}
+
+class RocksDBStateDataSourceV2ReadSuite extends StateDataSourceV2ReadSuite {
+  private var oldProviderClass: String = _
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+
+    oldProviderClass = spark.conf.get(SQLConf.STATE_STORE_PROVIDER_CLASS.key)
+    spark.conf.set(SQLConf.STATE_STORE_PROVIDER_CLASS.key,
+      classOf[RocksDBStateStoreProvider].getName)
+  }
+
+  override def afterAll(): Unit = {
+    if (oldProviderClass != null) {
+      spark.conf.set(SQLConf.STATE_STORE_PROVIDER_CLASS.key, oldProviderClass)
+    } else {
+      spark.conf.unset(SQLConf.STATE_STORE_PROVIDER_CLASS.key)
+    }
+
+    super.afterAll()
+  }
+}
+
+abstract class StateDataSourceV2ReadSuite extends StateDataSourceV2TestBase with Assertions {
+
+  test("simple aggregation, state ver 1") {
+    testStreamingAggregation(1)
+  }
+
+  test("simple aggregation, state ver 2") {
+    testStreamingAggregation(2)
+  }
+
+  test("composite key aggregation, state ver 1") {
```

Comment:
Can we add some test cases for failure conditions too ?

- not a stateful query
- schema is invalid
- options passed are invalid
- etc
?

Reply:
We seem to have bunch of callers to this function in a few places right ?

```
  private val keySchema = SchemaUtil.getSchemaAsDataType(stateSchema, "key")
    .asInstanceOf[StructType]
  private val valueSchema = SchemaUtil.getSchemaAsDataType(stateSchema, "value")
    .asInstanceOf[StructType]

  private val userFacingValueSchema = SchemaUtil.getSchemaAsDataType(userFacingSchema, "value")
    .asInstanceOf[StructType]
```

I wonder whether we need to test for the schema file here being invalid/corrupt is throwing an exception as expected ? just some negative cases, in case you think they are worth it/necessary ? if not, we can skip

Reply:
Yeah I'll check and come up with negative cases.

For the case of schema being invalid, there would be mostly no case because

* Users are unable to provide the schema as source option.
* If there is a state schema file, it is the source of truth. UnsafeRow (and underlying serialized data) does not have an information about schema and we just project the value based on the schema we have in state schema file. If there are mismatch between schema and the row, it would end up with runtime exception, and even worse, JVM level error. That was one of the main reason I broght up the state schema; previously there is no schema information at all and no check with compatibility.

But maybe I can revisit the case where the schema "file" is corrupted.